gensim/models/keyedvectors.py

```diff
@@ -559,6 +560,90 @@ def similar_by_vector(self, vector, topn=10, restrict_vocab=None):
+        dictionary : gensim.corpora.Dictionary
+            A dictionary associated with the corpus.
+        threshold : float, optional
+            Only pairs of words whose embeddings are more similar than `threshold` are considered
+            when building the sparse term similarity matrix. Defaults to `0.0`.
+        exponent : float, optional
+            The exponent applied to the similarity between two word embeddings when building the
+            term similarity matrix. Defaults to `2.0`.
+        nonzero_limit : int, optional
+            The maximum number of non-zero elements outside the diagonal in a single row or column
+            of the term similarity matrix. Setting `nonzero_limit` to a constant ensures that the
+            time complexity of computing the Soft Cosine Similarity will be linear in the document
+            length rather than quadratic. Defaults to `100`.
+        dtype : numpy.dtype, optional
+            Data-type of the term similarity matrix. Defaults to `numpy.float32`.
+
+        Returns
+        -------
+        scipy.sparse.csc_matrix
+            The constructed term similarity matrix.
+
+        See Also
+        --------
+        gensim.matutils.softcossim
+            The Soft Cosine Similarity.
+        gensim.similarities.docsim.SoftCosineSimilarity
+            A class for performing corpus-based similarity queries with Soft Cosine Similarity.
+
+        References
+        ----------
+        The constructed matrix corresponds to the matrix Mrel defined in section 2.1 of
+        [charletdamnati17]_.
+
+        .. [charletdamnati17] Delphine Charlet and Geraldine Damnati, "SimBow at SemEval-2017
+           Task 3: Soft-Cosine Semantic Similarity between Questions for Community Question
+           Answering", 2017.
+        """
+
+        logger.info("constructing a term similarity matrix")
+        similarity_matrix = sparse.identity(len(dictionary), dtype=dtype, format="lil")
+        # Traverse rows.
+        num_rows = len(dictionary)
+        for w1_index in range(num_rows):
+            if w1_index % ceil(num_rows / 10) == 0:
+                logger.info("PROGRESS: at row %i / %i", w1_index + 1, num_rows)
+            w1 = dictionary[w1_index]
+            if w1 not in self.vocab:
+                continue  # A word from the dictionary not present in the word2vec model.
+            # Traverse upper triangle columns.
+            if len(dictionary) <= nonzero_limit + 1:  # Traverse all columns.
+                columns = ((w2_index, self.similarity(w1, dictionary[w2_index]))
+                           for w2_index in range(w1_index + 1, num_rows)
+                           if w1_index != w2_index and dictionary[w2_index] in self.vocab)
+            else:  # Traverse only columns corresponding to the embeddings closest to w1.
+                num_nonzero = similarity_matrix[w1_index].getnnz() - 1
+                columns = ((dictionary.token2id[w2], similarity)
```

Comment:
Please use hanging indents (instead of vertical).

Reply:
Fixed in e1eb7cd12452fa48bf4dfb98ee33d67c4ea47ed0 and c7f0ce176ebda8843b725c673f8c6f950fdc85ce.